### PR TITLE
feat: semantic viewer new/edit/view modes

### PR DESCRIPTION
**Pull request by IrakliJani**
Closes: https://github.com/lightdash/lightdash/issues/11423

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/controllers/v2/SemanticLayerController.ts [hunk 1]
```diff
@@ -2,19 +2,24 @@ import {
     ApiErrorPayload,
```
Comment (IrakliJani) [inline packages/backend/src/controllers/v2/SemanticLayerController.ts:2]: any reason why I put this under v2?

File: packages/backend/src/controllers/v2/SemanticLayerController.ts [hunk 2]
```diff
@@ -194,8 +199,8 @@ export class SemanticLayerController extends BaseController {
     async createSemanticViewerChart(
```
Comment (IrakliJani) [inline packages/backend/src/controllers/v2/SemanticLayerController.ts:199]: methods from this one to below should be under `SavedSemanticViewerController.ts` instead

File: packages/frontend/src/features/semanticViewer/store/semanticViewerSlice.ts [hunk 3]
```diff
@@ -125,15 +124,9 @@ export interface SemanticViewerState {
     columns: VizColumn[];
     fields: SemanticLayerField[];
 
-    selectedDimensions: Record<string, SemanticLayerStateDimension>;
-    selectedTimeDimensions: Record<string, SemanticLayerStateTimeDimension>;
-    selectedMetrics: Record<string, SemanticLayerStateMetric>;
```
Comment (almeidabbm) [inline packages/frontend/src/features/semanticViewer/store/semanticViewerSlice.ts:130]: 🤌 🚀  Love this change, tracking selected fields via the `SemanticLayerQuery`

**Review by almeidabbm — APPROVED** [review]
Code looks super!! 🎉 
I just have a couple minor things, not sure if we want to fix them here or in separate PRs

**Comment by almeidabbm** [comment]
When trying to edit dashboard tile content (e.g. title) for Semantic Viewer chart, I get this (seems like the loom icon)

Screenshot 2024-09-25 at 14 33 25

**Comment by almeidabbm** [comment]
Couple places where URLs are incorrect (probably because we're using uuids here instead of slugs):
- Dashboard edit chart
- Dashboard view chart
- Opening chart from chart lists (e.g. spaces/home/all charts page)

**Comment by almeidabbm** [comment]
No longer can save a chart unless I click on `chart` tab, previous state allowed me to save a chart (as a table) as long as I had results, no need to click save on the chart tab

**Now**
image

**Before**
image

File: packages/frontend/src/features/semanticViewer/components/SaveChart/SemanticViewerSaveChartContent.tsx [hunk 4]
```diff
@@ -28,7 +44,48 @@ const SemanticViewerSaveChartContent: FC = () => {
...
+        selectedChartConfig,
+        semanticLayerQuery,
+        semanticLayerView,
+        showToastSuccess,
+    ]);
+
+    const canSave =
+        selectedFieldNames.length > 0 &&
+        !!selectedChartConfig &&
+        !!activeChartKind;
```
Comment (almeidabbm) [inline packages/frontend/src/features/semanticViewer/components/SaveChart/SemanticViewerSaveChartContent.tsx:88]: This is most likely the cause of the issue here: https://github.com/lightdash/lightdash/pull/11659#issuecomment-2374168083

We no longer require a activeChartKind to save, it should default to table

File: packages/frontend/src/pages/SemanticViewerEdit.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                    projectUuid,
+                    info: infoQuery.data,
+                    chart: chartQuery.data.chart,
+                }),
+            );
+        }
+
+        // If we don't have a saved chart, initialize the viewer with an empty chart
+        if (!savedSemanticViewerChartUuid && infoQuery.isSuccess) {
+            console.log('did this happen?');
```
Comment (almeidabbm) [inline packages/frontend/src/pages/SemanticViewerEdit.tsx:113]:
```suggestion
```

File: packages/frontend/src/pages/SemanticViewerEdit.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+    // TODO: add error state
+    if (infoQuery.isError || chartQuery.isError || fieldsQuery.isError) {
+        return null;
+    }
+
+    // TODO: add loading state
+    if (semanticViewerState === SemanticViewerStateStatus.LOADING) return null;
+
+    console.log(semanticViewerState);
```
Comment (almeidabbm) [inline packages/frontend/src/pages/SemanticViewerEdit.tsx:148]:
```suggestion
```

File: packages/frontend/src/features/semanticViewer/components/SaveChart/SemanticViewerSaveChartContent.tsx (hunk 4 above)
Comment (IrakliJani) [reply packages/frontend/src/features/semanticViewer/components/SaveChart/SemanticViewerSaveChartContent.tsx:88]: ah, makes sense!

**Comment by almeidabbm** [comment]
When I save a table viz it won't display anything when going to the view page:

https://github.com/user-attachments/assets/c1e034bb-11f8-478e-9623-1596214def52

File: packages/frontend/src/features/semanticViewer/api/hooks.ts [hunk 7]
```diff
@@ -133,95 +137,122 @@ export const useSemanticLayerQueryResults = (
...
+
+            // If the chart is not a table, we don't need to fetch the results.
+            if (chart.config.type !== ChartKind.TABLE) {
+                return {
+                    chart,
+                    // TODO: this feels like a type hack
+                    resultsAndColumns: {
+                        results: [],
+                        columns: [],
+                    },
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/frontend/src/features/semanticViewer/api/hooks.ts]: it did not make any sense

**Comment by IrakliJani** [comment]
> When trying to edit dashboard tile content (e.g. title) for Semantic Viewer chart, I get this (seems like the loom icon)
> 
> Screenshot 2024-09-25 at 14 33 25

I'll make sure to create a separate ticket

**Comment by IrakliJani** [comment]
> When I save a table viz it won't display anything when going to the view page:
> Screen.Recording.2024-09-25.at.15.20.39.mov

uuid url's are going to be a separate PR

**Review by almeidabbm — APPROVED** [review]
LGTM 🚀

**Pull request merged**